src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs

```diff
@@ -287,186 +216,323 @@ void ProcessStack ()
+				Collection<Instruction> instrs = body.Instructions;
+
+				for (int i = 0; i < body.Instructions.Count; ++i) {
+					Instruction instr = instrs[i];
+					switch (instr.OpCode.Code) {
+
+					case Code.Call:
+					case Code.Callvirt:
+						MethodDefinition? md = optimizer._context.TryResolve ((MethodReference) instr.Operand);
+						if (md == null)
+							continue;
+
+						if (md.IsVirtual)
+							continue;
+
+						if (md.CallingConvention == MethodCallingConvention.VarArg)
 							break;
-						reducer.Rewrite (i, targetResult);
-						changed = true;
-					}
-					break;
 
-				case Code.Sizeof:
-					//
-					// sizeof (IntPtr) and sizeof (UIntPtr) are just aliases for IntPtr.Size and UIntPtr.Size
-					// which are simple static properties commonly overwritten. Instead of forcing C# code style
-					// we handle both via static Size property
-					//
-					MethodDefinition? sizeOfImpl = null;
+						if (md.NoInlining)
+							break;
 
-					var operand = (TypeReference) instr.Operand;
-					if (operand.MetadataType == MetadataType.UIntPtr) {
-						sizeOfImpl = (UIntPtrSize ??= FindSizeMethod (_context.TryResolve (operand)));
-					} else if (operand.MetadataType == MetadataType.IntPtr) {
-						sizeOfImpl = (IntPtrSize ??= FindSizeMethod (_context.TryResolve (operand)));
-					}
+						var cpl = new CalleePayload (md, GetArgumentsOnStack (md, body.Instructions, i));
+						MethodResult? call_result = optimizer.TryGetMethodCallResult (cpl, optimizer.GetRecursionGuard ());
+						if (call_result is not MethodResult result)
+							break;
 
-					if (sizeOfImpl != null) {
-						if (!TryGetConstantResultForMethod (sizeOfImpl, out targetResult)) {
-							if (!treatUnprocessedDependenciesAsNonConst)
-								hasUnprocessedDependencies = true;
+						if (!result.IsSideEffectFree)
 							break;
-						} else if (targetResult == null || hasUnprocessedDependencies) {
+
+						TypeDefinition methodType = md.DeclaringType;
+						if (methodType != body.Method.DeclaringType && !methodType.IsBeforeFieldInit) {
+							//
+							// Figuring out at this point if the explicit static ctor will be used is hard
+							//
+							optimizer._context.LogMessage ($"Cannot inline result of '{md.GetDisplayName ()}' call due to presence of static constructor");
 							break;
 						}
 
-						reducer.Rewrite (i, targetResult);
+						if (!md.IsStatic) {
+							if (!md.HasParameters && CanInlineInstanceCall (instrs, i)) {
+								processor.Replace (i - 1, Instruction.Create (OpCodes.Nop));
+								processor.Replace (i, result.GetPrototype ()!);
+								changed = true;
+							}
+
+							continue;
+						}
+
+						if (md.HasParameters) {
+							if (!IsCalledWithoutSideEffects (md, instrs, i))
+								continue;
+
+							for (int p = 1; p <= md.Parameters.Count; ++p) {
+								processor.Replace (i - p, Instruction.Create (OpCodes.Nop));
+							}
+						}
+
+						processor.Replace (i, result.GetPrototype ());
```

Comment:
It can be both, shallow copy for complex instruction, deep copy for simple instruction. I have upstream PR which I'll merge here once approved (including the name)